### PR TITLE
Sortable header for component list in web UI

### PR DESCRIPTION
**Pull request by hainenber**
#### Which issue(s) this PR fixes

Implements #5616 

#### Notes to the Reviewer
Please check if the screen recording is as expected or something more formal needs to be done, i.e. tests

https://github.com/grafana/agent/assets/41283691/b037e1e4-efa2-47ad-b6f2-89012f3a6652

#### PR Checklist

- [x] CHANGELOG.md updated

### CODE REVIEW

File: web/ui/src/features/component/TableHead.tsx
Comment (wildum) [inline web/ui/src/features/component/TableHead.tsx:10]: I think that the order could be either a bool that you could just toggle with setOrder(!order). But if you want to keep an explicit state then I would suggest using an enum rather than a string

File: web/ui/src/features/component/TableHead.tsx
Comment (wildum) [inline web/ui/src/features/component/TableHead.tsx:16]: the setters should be inside of the if statement because if handleSorting is undefined then there is not point of changing the value of sortField and order.

File: web/ui/src/pages/PageComponentList.tsx
Comment (wildum) [inline web/ui/src/pages/PageComponentList.tsx:18]:
```suggestion
```
this line can be removed because you are checking the null values above already

File: web/ui/src/pages/PageComponentList.tsx
Comment (wildum) [inline web/ui/src/pages/PageComponentList.tsx:12]: to avoid nesting you could have if (!sortfield) return;

File: web/ui/src/pages/PageComponentList.tsx
Comment (wildum) [inline web/ui/src/pages/PageComponentList.tsx:25]: maybe you could have the check also at the beginning. so it woud then be if (!sortfield || !sortOrder) return;

**Review by wildum — CHANGES_REQUESTED** [review]
Thanks once again for your contribution! It's nice to see improvements for the UI and the little video that you added in description is very helpful. My knowledge of react is very limited but I left some suggestions

**Comment by hainenber** [comment]
hi there, all of your proposals are implemented, I wholeheartedly agreed with all of them. The added changes don't change the sorting logic and visuals, so the video still stays valid. PTAL if you have time again. Thank you!

File: web/ui/src/features/component/Table.module.css
Comment (wildum) [inline web/ui/src/features/component/Table.module.css:56]: is "box-shadow: none !important;" needed here?
can we simplify the CSS with:

.table tr th {
  cursor: pointer;
}

.table tr th[data-sort-order="desc"]:before {
  content: "▼";
  float: right;
}

.table tr th[data-sort-order="asc"]:before {
  content: "▲";
  float: right;
}

File: web/ui/src/features/component/TableHead.tsx
Comment (wildum) [inline web/ui/src/features/component/TableHead.tsx:15]: I found the logic a little bit hard to understand. Maybe we could refactor the code a bit to make the two following scenarios more obvious:
- the user clicks on a new header
- the user clicks on the same header another time

I would suggest this, although it is a bit verbose it clearly handles both scenarios:
```
if (header === sortField) {
  // If the user clicked again on the same header we toggle the sort order 
  sortOrder = order === SortOrder.ASC ? SortOrder.DESC : SortOrder.ASC;
} else {
  // If the user clicks on a new header we sort the entries by default in ASC order
  sortOrder = SortOrder.ASC;
}
```

File: web/ui/src/features/component/TableHead.tsx
Comment (wildum) [inline web/ui/src/features/component/TableHead.tsx:31]: can we change it to:
`data-sort-order={sortField === header ? order : undefined}`
and get rid of `const sortOrderHeaderAttribute = sortField === header ? order : 'default';` ?

File: web/ui/src/features/component/Table.module.css
Comment (wildum) [reply web/ui/src/features/component/Table.module.css:56]: if you remove the 'default' as stated in my other comment then I believe that this should be ok

File: web/ui/src/pages/PageComponentList.tsx
Comment (wildum) [inline web/ui/src/pages/PageComponentList.tsx:26]: To make the sorting easier to extend you could use a map:

```
const fieldMappings: { [key: string]: (comp: any) => string | undefined } = {
    Health: comp => comp.health?.state?.toString(),
    localID: comp => comp.localID,
   // Add new fields if needed here.
};

function getSortValue(component: any, field: string): string | undefined {
    const valueGetter = fieldMappings[field];
    return valueGetter ? valueGetter(component) : undefined;
}
```

And you could use it as follow in the sorting function:
(sortValueA and sortValueB can now be 'undefined', that's why I changed the if statement as well)
```
    const sortValueA = getSortValue(a, sortField);
    const sortValueB = getSortValue(b, sortField);

    if (!sortValueA) return 1;
    if (!sortValueB) return -1;
```

**Review by wildum — CHANGES_REQUESTED** [review]
Thanks for adapting the code, I did a second round of review with a few more suggestions

**Comment by wildum** [comment]
One more thing, after discussing with our new UX designer, we would like to have by default the sorting enabled with ID "asc". This way the arrow will already be shown and it's easier for the user to find the sorting functionality

File: web/ui/src/features/component/Table.module.css
Comment (hainenber) [reply web/ui/src/features/component/Table.module.css:52]: CSS block for `default` table header removed. I believe this can be resolved now.

File: web/ui/src/features/component/TableHead.tsx
Comment (hainenber) [reply web/ui/src/features/component/TableHead.tsx]: Done this one too in latest commit.

File: web/ui/src/pages/PageComponentList.tsx
Comment (hainenber) [reply web/ui/src/pages/PageComponentList.tsx:37]: Done this one too in latest commit.

File: web/ui/src/features/component/TableHead.tsx
Comment (hainenber) [reply web/ui/src/features/component/TableHead.tsx]: Done this one too in latest commit.

**Comment by hainenber** [comment]
Hey, thanks for another review. I've made the refactor to simplify the sorting logic here and there as well as making ID column to be sorted ascendingly by default. PTAL again

File: CHANGELOG.md
Comment (wildum) [inline CHANGELOG.md:121]: duplicated entries to remove

File: web/ui/src/features/component/TableHead.tsx
Comment (wildum) [inline web/ui/src/features/component/TableHead.tsx:18]:
```suggestion
    // User clicks again on the header, we toggle the previous sort order
```

**Review by wildum — CHANGES_REQUESTED** [review]
Thanks for making the changes. I added two little comments but it looks good, nice job :)

File: web/ui/src/pages/PageComponentList.tsx
Comment (wildum) [inline web/ui/src/pages/PageComponentList.tsx:14]: nit: we actually should replace any (+ the one ligne 8) with Component. This was my mistake in the code I suggested and was caught by the linter: " Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any"

Comment (hainenber) [reply web/ui/src/pages/PageComponentList.tsx]: Fixed in latest commit

File: CHANGELOG.md
Comment (hainenber) [reply CHANGELOG.md]: Duplicate entry removed. Thanks for pointing this one out!

File: CHANGELOG.md
Comment (wildum) [inline CHANGELOG.md:119]: this entry is still duplicated

Comment (hainenber) [reply CHANGELOG.md]: Removed in latest commit

**Review by wildum — APPROVED** [review]
Good work, it looks good to me, you will still need to update with main because the changelog has conflicts. I'm not an official maintainer yet so I can't merge it myself. I will ping my colleagues to take over from here

**Pull request merged**
